### PR TITLE
Allow blast radius projectiles to explode when their lifetime runs out

### PR DESCRIPTION
**Pull request by Amazinite**
**Feature**

This PR addresses the feature described in issue #9765.

## Summary
Add a new weapon attribute with no value (like "cluster" or "safe").
* `"fused"`: Causes the projectile to explode when it reaches the last frame of its lifetime. Normally, a projectile that reaches the last frame of its lifetime just disappears, creating any "die effect" that may be present on the weapon.

## Usage examples
The following weapon would create an explosion the moment it is fired, dealing 1000 shield or hull damage to everything within 1000 units. The use of "fused" is what causes the projectile to instantly create an explosion without the need of a trigger radius. The use of "safe range override" allows NPCs to ignore the fact that they're about to also damage themself. This is effectively what is desired out of an aura weapon as described in the linked issue.
```
weapon
	reload 180
	lifetime 1
	velocity 0
	"range override" 1000
	"blast radius" 1000
	"safe range override" 0
	"shield damage" 1000
	"hull damage" 1000
	"fused"
```

## Testing Done
Added `"fused"` to a heavy rocket launcher. Without this PR, heavy rockets that reach the end of their lifetime create a small explosion effect, but don't actually damage anything nearby. With this PR, the rockets create their normal hit effect at the end of their lifetime and cause damage to nearby ships.

### CODE REVIEW

File: source/Projectile.cpp [hunk 1]
```diff
@@ -463,3 +463,10 @@ uint16_t Projectile::HitsRemaining() const
 {
 	return hitsRemaining;
 }
+
+
+
+bool Projectile::ShouldExplode() const
+{
+	return !government || weapon->IsExplodeOnDeath() && lifeitme == 1;
```
Comment (Amazinite) [inline source/Projectile.cpp:471]:
```suggestion
	return !government || (weapon->IsExplodeOnDeath() && lifetime == 1);
```

File: source/Weapon.h [hunk 2]
```diff
@@ -420,6 +425,7 @@ inline bool Weapon::IsSafe() const { return isSafe; }
 inline bool Weapon::IsPhasing() const { return isPhasing; }
 inline bool Weapon::IsDamageScaled() const { return isDamageScaled; }
 inline bool Weapon::IsGravitational() const { return isGravitational; }
+inline bool Weapon::IsExplodeOnDeath() const { return isExplodeOnDeath; }
```
Comment (Hurleveur) [inline source/Weapon.h:428]: I'd rename this to ExplodesOnDeath() honestly, we dont have IsBlastRadius

**Review by Hurleveur — COMMENTED** [review]
I'm a little bit annoyed it explodes a lifetime 1 instead of 0 but looking at the code it's either that or we'd have to delay its destruction another step which would be less pretty code ig.

So looks good to me, we could always override a lifetime 0 implicit projectile like this to 1, but we dont have to.

**Comment by Hurleveur** [comment]
What about getting shot down by AM actually?
I dont think that would trigger this, nor should it but we should mb plan a keyword for that as well
Because explodes on death kinda sounds like smth AM should trigger

File: source/Weapon.h (hunk 2 above)
Comment (Amazinite) [reply source/Weapon.h:428]: Well blast radius is a double, not a boolean.

**Comment by Amazinite** [comment]
> A less verbose name, like `fused` for example might work a little better?

> What about getting shot down by AM actually?

I think that "fused" makes sense if we don't also add exploding upon being hit by AM to this. We can look at exploding on AM death later.

**Comment by Amazinite** [comment]
> I'm a little bit annoyed it explodes a lifetime 1 instead of 0 but looking at the code it's either that or we'd have to delay its destruction another step which would be less pretty code ig.

The frame where a projectile is at lifetime 1 is the last frame that the projectile is actually able to collide anyway. The way that Engine works is that projectiles are moved (which decrements their lifetime counter), dead projectiles are pruned (a projectile with 0 lifetime is dead), and then collision detection is run on the remaining projectiles. So having the projectile explode on lifetime 0 would actually effectively be increasing the lifetime of the projectile by 1. The way this behave now is like if a trigger radius projectile had something enter its trigger radius on the very last frame that it runs collision detection.

**Review by Hurleveur — COMMENTED** [review]
Shouldn't we add a tooltip or smth for this? Or is it meant to be invisible to the player?

Code looks good to me, we'll just need to be careful about specifying you need a lifetime of 1 in the description.

**Comment by Amazinite** [comment]
This isn't displayed in the outfitter, as neither are cluster, stream, safe, etc. 
The only boolean attributes that are displayed are these ones.
https://github.com/endless-sky/endless-sky/blob/8ed9ec0287a48b4715a9338f965bb6867222bc78/source/OutfitInfoDisplay.cpp#L208-L216

**Review by Hurleveur — APPROVED** [review]
Seems fair, could be added in another PR but the code for this looks good

**Pull request merged**